flipt-engine-wasm/src/lib.rs

```diff
@@ -51,145 +68,286 @@ where
+        let payload =
+            std::str::from_utf8_unchecked(std::slice::from_raw_parts(payload_ptr, payload_len));
+
+        match Engine::new(namespace, payload) {
+            Ok(engine) => Box::into_raw(Box::new(engine)) as *mut c_void,
+            Err(e) => {
+                eprintln!("Error initializing engine: {}", e);
+                std::ptr::null_mut()
+            }
+        }
+    });
+
+    result.unwrap_or(std::ptr::null_mut())
+}
+
+/// # Safety
+///
+/// This function will take in a pointer to the engine and return a variant evaluation response.
+#[no_mangle]
+pub unsafe extern "C" fn evaluate_variant(
+    engine_ptr: *mut c_void,
+    evaluation_request_ptr: *const u8,
+    evaluation_request_len: usize,
+) -> u64 {
+    let e = get_engine(engine_ptr).unwrap();
+    let evaluation_request = unsafe {
+        std::str::from_utf8_unchecked(std::slice::from_raw_parts(
+            evaluation_request_ptr,
+            evaluation_request_len,
+        ))
+    };
+
+    let request = get_evaluation_request(evaluation_request);
+    let result = result_to_string(e.evaluate_variant(&request));
+    let (ptr, len) = string_to_ptr(&result);
+    std::mem::forget(result); // host owns the string now and must deallocate it
+    ((ptr as u64) << 32) | len as u64 // return u64 with ptr and len as high and low bits
+}
+
+/// # Safety
+///
+/// This function will take in a pointer to the engine and return a boolean evaluation response.
+#[no_mangle]
+pub unsafe extern "C" fn evaluate_boolean(
+    engine_ptr: *mut c_void,
+    evaluation_request_ptr: *const u8,
+    evaluation_request_len: usize,
+) -> u64 {
+    let e = get_engine(engine_ptr).unwrap();
+    let evaluation_request = unsafe {
+        std::str::from_utf8_unchecked(std::slice::from_raw_parts(
+            evaluation_request_ptr,
+            evaluation_request_len,
+        ))
+    };
+
+    let request = get_evaluation_request(evaluation_request);
+    let result = result_to_string(e.evaluate_boolean(&request));
+    let (ptr, len) = string_to_ptr(&result);
+    std::mem::forget(result); // host owns the string now and must deallocate it
+    ((ptr as u64) << 32) | len as u64 // return u64 with ptr and len as high and low bits
+}
 
-    pub fn evaluate_variant(&self, request: JsValue) -> Result<JsValue, JsValue> {
-        let result: Result<fliptevaluation::VariantEvaluationResponse, Error> =
-            match serde_wasm_bindgen::from_value(request) {
-                Ok(req) => variant_evaluation(&self.store, &self.namespace, &req),
-                Err(e) => Err(Error::InvalidRequest(e.to_string())),
-            };
+/// # Safety
+///
+/// This function will take in a pointer to the engine and return a batch evaluation response.
+#[no_mangle]
+pub unsafe extern "C" fn evaluate_batch(
+    engine_ptr: *mut c_void,
+    batch_evaluation_request_ptr: *const u8,
+    batch_evaluation_request_len: usize,
+) -> u64 {
+    let e = get_engine(engine_ptr).unwrap();
```

Comment:
It would be nice not to panic on the error and just return the error response to the caller.

Reply:
Good shout! This is now handled by the new `result_to_ptr` funcs

Reply:
I also handled this in the FFI lib as well

Reply:
very nice!